### PR TITLE
[ADR] JAMES-3400 James CLI based on webadmin endpoints 

### PR DESCRIPTION
**Pull request by quantranhong1999**
ADR for James CLI based on webadmin endpoints.

### CODE REVIEW

**Review by chibenwa — COMMENTED** [review]
Can you just keep the changes concerning the ADR here, using git rebase ?

Also stash them all in a single commit!

Thanks in advance ;-)

**Comment by chibenwa** [comment]
Please can you add the ticket number as a prefix?

`JAMES-3400 the rest of the message`

You can do that with rebase + reword

File: src/adr/0042-james-cli-based-on-webadmin.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
+# 42. James CLI based on webadmin API
+Date: 2020-10-05
```
Comment (Arsnael) [inline src/adr/0042-james-cli-based-on-webadmin.md:2]: Skip a line between the date and title, for readability

File: src/adr/0042-james-cli-based-on-webadmin.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+## Status
+Accepted (lazy consensus).
+
+ ## Context
```
Comment (Arsnael) [inline src/adr/0042-james-cli-based-on-webadmin.md:7]: Remove that extra space before `##`

File: src/adr/0042-james-cli-based-on-webadmin.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+## Decision
+
+We decided to write a new CLI client, running on top of the JVM, communicating with James via the webadmin protocol, using http.
+
+* What libraries will we use? 
+
+  * http client: ***Feign library***. We used it as an http client in other parts of James so we continue to use it.
+
+  * CLI: ***Picocli library***. Picocli is a one-file command line parsing framework writed by Java that allows we to create command line applications with almost no code. It allows mixing Options with positional Parameters (Eg: don't have to follow order Options then Parameters), [automatic type conversion](https://picocli.info/#_strongly_typed_everything) of command line arguments to the type of the annotated field, provide Automatic Help and better Subcommand Support, easily handle Exceptions.
```
Comment (Arsnael) [inline src/adr/0042-james-cli-based-on-webadmin.md:25]:
```suggestion
  * CLI: ***Picocli library***. Picocli is a one-file command line parsing framework written in Java that allows us to create command line applications with almost no code. It allows mixing Options with positional Parameters (Eg: no need to follow the order Options then Parameters), [automatic type conversion](https://picocli.info/#_strongly_typed_everything) of command line arguments to the type of the annotated field, provide Automatic Help and better Subcommand Support, easily handle Exceptions.
```

**Review by Arsnael — COMMENTED** [review]
Overall nice job :)

**Review by chibenwa — APPROVED** [review]

**Review by mbaechler — APPROVED** [review]
This is a good thing to finally switch admin to http.

However we should drop legacy things like jar+lib packaging, making people running `java -jar ...`, etc

Also, I would add:
* write a man page (asciidoc can generate it for you)
* provide the right tool to enable shell completion

But I would already be happy if none of my comments were adopted.

File: src/adr/0042-james-cli-based-on-webadmin.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+* How will we limit breaking changes this new CLI will cause?
+
+  * Work on a wrapper to adapt the old CLI API.
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply src/adr/0042-james-cli-based-on-webadmin.md:31]: credits to @rouazana  :-)

File: src/adr/0042-james-cli-based-on-webadmin.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+* Where will we locate this cli code?
+
+  * server/protocols/webadmin-cli
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply src/adr/0042-james-cli-based-on-webadmin.md:35]: IMO that's a promotion that could come later

File: src/adr/0042-james-cli-based-on-webadmin.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+* We decided to adopt a more modern, modular CLI syntax:
+
+```   
+$ java -jar james_cli.jar [OPTION] ENTITY ACTION {ARGUMENT}
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply src/adr/0042-james-cli-based-on-webadmin.md:40]: What do you propose instead?

File: src/adr/0042-james-cli-based-on-webadmin.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+Webadmin APIs use HTTP protocol, which is more secure than JMX protocol to interact with James servers.
+
+Webadmin command-line interface is an upcoming replacement for the outdated, security-vulnerable JMX command-line interface. 
+
+## Decision
+
+We decided to write a new CLI client, running on top of the JVM, communicating with James via the webadmin protocol, using http.
+
+* What libraries will we use?
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply src/adr/0042-james-cli-based-on-webadmin.md:23]: Later ADRs can superseed this and overide the original choice?

File: src/adr/0042-james-cli-based-on-webadmin.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+Besides, the current CLI interface is also not optimal for users. It places actions in front of entities with contiguous syntax, making it harder for the user to remember the command (for example, which entity the GET action command can interact with). If we design to place the entity first and the outgoing actions can interact with that entity afterward, the user will easily imagine what he/she can do with each entity. This creates an intuitive interface that is easier to remember.
+
+Webadmin APIs use HTTP protocol, which is more secure than JMX protocol to interact with James servers.
+
+Webadmin command-line interface is an upcoming replacement for the outdated, security-vulnerable JMX command-line interface. 
+
+## Decision
+
+We decided to write a new CLI client, running on top of the JVM, communicating with James via the webadmin protocol, using http.
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply src/adr/0042-james-cli-based-on-webadmin.md:21]: Agree with distributing this CLI as a GraalVM native image :+1: 

Looks easily achievable https://www.graalvm.org/reference-manual/native-image/NativeImageMavenPlugin/

File: src/adr/0042-james-cli-based-on-webadmin.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (chibenwa) [reply src/adr/0042-james-cli-based-on-webadmin.md:40]: Ok, graalVM allows to avoid that

**Comment by chibenwa** [comment]
> write a man page (asciidoc can generate it for you)

Picocli generates beautiful documentation for your application (HTML, PDF and Unix man pages)

Will be easy, let's add the mention!

File: server/protocols/webadmin-cli/README.md [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+Webadmin command-line interface is an upcoming replacement for the outdated, security-vulnerable JMX command-line interface. It also aims at providing a more modern and intuitive interface.
+
+## Run the command line
+
+## Syntax
+
+General syntax to run the command line
+
+```   
+$ java -jar james_cli.jar [OPTION] ENTITY ACTION {ARGUMENT}
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/README.md:14]: Replace `java -jar james_cli.jar` by `james-cli`

File: src/adr/0042-james-cli-based-on-webadmin.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (chibenwa) [reply src/adr/0042-james-cli-based-on-webadmin.md:46]: Quan can you replace `java -jar james_cli.jar` by `james-cli` ?

File: src/adr/0042-james-cli-based-on-webadmin.md [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+  
+    ACTION: name of the action to perform,
+  
+    ARGUMENT: arguments needed for the action.
+
+#### Examples
+
+Add a domain to the domain list.
+```
+$ java -jar james_cli.jar --url http://127.0.0.1 --port 9999 domain create domainNameToBeCreated
```
Comment (chibenwa) [inline src/adr/0042-james-cli-based-on-webadmin.md:62]: Here too, replace `java -jar james_cli.jar` by `james-cli`

File: src/adr/0042-james-cli-based-on-webadmin.md [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+
+## Consequences
+
+It aims at providing a more modern and more secure CLI, also bringing compatibility ability with old CLI.
+
+## References
+* [NVD-CVE-2017-12628 Detail](https://nvd.nist.gov/vuln/detail/CVE-2017-12628)
+* [Picocli 2.0: Do More With Less](https://dzone.com/articles/whats-new-in-picocli-20)
+* [Picocli Homepage](https://picocli.info/)
```
Comment (chibenwa) [inline src/adr/0042-james-cli-based-on-webadmin.md:83]: Can we add this reference? https://www.graalvm.org/reference-manual/native-image/NativeImageMavenPlugin/

File: server/protocols/webadmin-cli/README.md (hunk 9 above)
Comment (chibenwa) [inline server/protocols/webadmin-cli/README.md:14]: Can we replace `java -jar james_cli.jar` by `james-cli` ?

File: server/protocols/webadmin-cli/README.md [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+  
+    ENTITY: represents the entity to perform action on,
+  
+    ACTION: name of the action to perform,
+  
+    ARGUMENT: arguments needed for the action.
+
+Example: 
+```
+$ java -jar james_cli.jar --url http://127.0.0.1 --port 9999 domain list
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/README.md:29]: Can we replace `java -jar james_cli.jar` by `james-cli` ?

File: server/protocols/webadmin-cli/README.md [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+```
+
+The above command lists all domain names available on domain route at address http://127.0.0.1:9999. 
+It does not require any argument to execute. Options --url and --port are optional. Without them, the default value is http://127.0.0.0:8000.
+As for other commands, arguments might be required after the sub-command (ACTION such as list, add and remove).
+
+Note: the command line ```$ java -jar james_cli.jar``` will be shorten as {cli} from here.
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/README.md:36]: Remove this note

File: server/protocols/webadmin-cli/README.md [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+   - [Delete an user](#delete-a-user)
+   - [Get users list](#get-users-list)
+   - [Update an user password](#update-a-user-password)
+
+## Manage Domains
+
+### Create a domain
+Add a domain to the domain list.
+```
+{cli} domain create <domainToBeCreated>
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/README.md:57]: Can we replace `{cli}` by `james-cli` ?

File: server/protocols/webadmin-cli/README.md [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+- can not be null or empty
+- can not contain ‘@’
+- can not be more than 255 characters
+- can not contain ‘/’
+
+### Delete a domain
+
+Remove a domain from the domain list.
+```
+{cli} domain delete <domainToBeDeleted>
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/README.md:70]: Can we replace `{cli}` by `james-cli` ?

File: server/protocols/webadmin-cli/README.md [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+```
+Note: Deletion of an auto-detected domain, default domain or of an auto-detected ip is not supported. We encourage you instead to review your [domain list configuration](https://james.apache.org/server/config-domainlist.html).
+
+### Check if a domain exists
+Check whether a domain exists on the domain list or not.
+```
+{cli} domain exist <domainToBeChecked>
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/README.md:77]: Can we replace `{cli}` by `james-cli` ?

File: server/protocols/webadmin-cli/README.md [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+```
+
+### Get the list of domains
+Show all domains' name on the list.
+```
+{cli} domain list
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/README.md:83]: Can we replace `{cli}` by `james-cli` ?

File: server/protocols/webadmin-cli/README.md [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+
+
+
+## Manage Users
+
+### Create an user
+
+Add an user to the user list.
+```
+{cli} user create <username> <password>
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/README.md:94]: Can we replace `{cli}` by `james-cli` ?

File: server/protocols/webadmin-cli/README.md [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+```
+Resource name <username> representing valid users, hence it should match the criteria at [User Repositories documentation](https://james.apache.org/server/config-users.html)
+
+Note: if the user exists already, its password will be updated.
+
+### Test an user existence
+
+Check whether an user exists on the user list or not.
+```
+{cli} user exist <username>
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/README.md:104]: Can we replace `{cli}` by `james-cli` ?

File: server/protocols/webadmin-cli/README.md [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+```
+Resource name <username> representing valid users, hence it should match the criteria at [User Repositories documentation](https://james.apache.org/server/config-users.html)
+
+### Delete an user
+
+Remove an user from the user list.
+```
+{cli} user delete <username>
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/README.md:112]: Can we replace `{cli}` by `james-cli` ?

File: server/protocols/webadmin-cli/README.md [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+```
+
+### Get users list
+
+Show all users' name on the list.
+
+```
+{cli} user list
```
Comment (chibenwa) [inline server/protocols/webadmin-cli/README.md:120]: Can we replace `{cli}` by `james-cli` ?

**Review by chibenwa — COMMENTED** [review]
Almost there!

Let's update the README as well so that I can merge it all at once ;-)

File: src/adr/0042-james-cli-based-on-webadmin.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (quantranhong1999) [reply src/adr/0042-james-cli-based-on-webadmin.md]: done it

**Review by chibenwa — APPROVED** [review]

File: src/adr/0042-james-cli-based-on-webadmin.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (mbaechler) [reply src/adr/0042-james-cli-based-on-webadmin.md:35]: could you argue about putting it inside server/protocols? I can't find any rational for that.

File: src/adr/0042-james-cli-based-on-webadmin.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (mbaechler) [reply src/adr/0042-james-cli-based-on-webadmin.md:23]: It's just too many details, as far as I know we never wrote an ADR about our libraries choices.

File: src/adr/0042-james-cli-based-on-webadmin.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (chibenwa) [reply src/adr/0042-james-cli-based-on-webadmin.md:35]: It is related to webadmin, having it close to webadmin feels natural to me.

Putting something that is not (yet) the official cli in an official cli place might be confusing. I would defer that choice.

**Comment by chibenwa** [comment]
Merged

Quan, can you close this?

**Pull request closed without merging**